File: python/lib/state_manager.py
#!/usr/bin/env python3
"""
State management for the Issue Orchestrator.

This module handles persistent state storage and retrieval for issue processing.
"""

import json
import logging
import sys
from pathlib import Path
from typing import Dict, List, Optional

try:
    import aiofiles
    HAS_AIOFILES = True
except ImportError:
    HAS_AIOFILES = False
    logging.warning("aiofiles not installed. Using synchronous file operations.")

from .models import IssueState


class StateManager:
    """Manages persistent state of issue processing"""
    
    def __init__(self, state_file: str = "processing-state.json"):
        # Always use absolute path to ensure state file is in the orchestrator directory
        if Path(state_file).is_absolute():
            self.state_file = Path(state_file)
        else:
            # Use the script's directory as the base for relative paths
            script_dir = Path(__file__).parent.parent.resolve()
            self.state_file = script_dir / state_file
        
        self.states: Dict[int, IssueState] = {}
        self._load_states()
    
    def _load_states(self):
        """Load states from file"""
        if self.state_file.exists():
            try:
                with open(self.state_file) as f:
                    data = json.load(f)
                    for issue_num, state_data in data.items():
                        self.states[int(issue_num)] = IssueState(**state_data)
            except (json.JSONDecodeError, TypeError) as e:
                logging.error(f"Error loading state file: {e}")
                self.states = {}
    
    async def save_states(self):
        """Save states to file"""
        data = {
            str(issue_num): state.to_dict() 
            for issue_num, state in self.states.items()
        }
        
        if HAS_AIOFILES:
            async with aiofiles.open(self.state_file, 'w') as f:
                await f.write(json.dumps(data, indent=2))
                await f.flush()  # Ensure data is written to disk
        else:
            # Fallback to synchronous file operations
            with open(self.state_file, 'w') as f:
                json.dump(data, f, indent=2)
                f.flush()
    
    def get_state(self, issue_number: int) -> Optional[IssueState]:
        """Get state for an issue"""
        return self.states.get(issue_number)
    
    def set_state(self, issue_number: int, state: IssueState):
        """Set state for an issue"""
        self.states[issue_number] = state
        
    def remove_state(self, issue_number: int):
        """Remove state for an issue"""
        if issue_number in self.states:
            del self.states[issue_number]
    
    def get_active_issues(self) -> List[int]:
        """Get list of issues currently being processed"""
        return [
            issue_num for issue_num, state in self.states.items()
            if state.status in ["running", "needs_input"]
        ]
    
    def get_available_agent_index(self, max_concurrent: int) -> Optional[int]:
        """Get the next available agent index"""
        used_agents = set()
        for state in self.states.values():
            if state.status in ["running", "needs_input"]:
                if state.agent_index is not None:
                    used_agents.add(state.agent_index)
        
        # Find first available agent index
        for i in range(max_concurrent):
            if i not in used_agents:
                return i
        
        return None
    
    def get_agent_issues(self, agent_index: int) -> List[int]:
        """Get list of issues being processed by a specific agent"""
        return [
            issue_num for issue_num, state in self.states.items()
            if state.agent_index == agent_index and 
            state.status in ["running", "needs_input"]
        ]

File: src/lib/stateManager.ts
import { mkdir, readFile, writeFile } from "fs/promises";
import { dirname, resolve } from "path";
import { IssueState, ProcessStatus } from "./models";

function resolveStatePath(stateFile: string): string {
  if (stateFile.startsWith("/")) {
    return stateFile;
  }
  return resolve(process.cwd(), stateFile);
}

export class StateManager {
  private readonly stateFile: string;
  private states: Map<number, IssueState> = new Map();

  constructor(stateFile = "processing-state.json") {
    this.stateFile = resolveStatePath(stateFile);
  }

  async initialize(): Promise<void> {
    try {
      const content = await readFile(this.stateFile, "utf8");
      const raw = JSON.parse(content) as Record<string, any>;
      for (const [key, value] of Object.entries(raw)) {
        const issueNumber = Number(key);
        if (!Number.isNaN(issueNumber)) {
          this.states.set(issueNumber, IssueState.fromJSON({
            issue_number: issueNumber,
            ...value,
          }));
        }
      }
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== "ENOENT") {
        console.error("Failed to load state file", error);
      }
    }
  }

  getState(issueNumber: number): IssueState | undefined {
    return this.states.get(issueNumber);
  }

  setState(issueNumber: number, state: IssueState): void {
    this.states.set(issueNumber, state);
  }

  removeState(issueNumber: number): void {
    this.states.delete(issueNumber);
  }

  async saveStates(): Promise<void> {
    const json: Record<string, any> = {};
    for (const [issueNumber, state] of this.states.entries()) {
      json[String(issueNumber)] = state.toJSON();
    }

    await mkdir(dirname(this.stateFile), { recursive: true });
    await writeFile(this.stateFile, JSON.stringify(json, null, 2), "utf8");
  }

  getActiveIssues(): number[] {
    return Array.from(this.states.entries())
      .filter(([, state]) =>
        state.status === ProcessStatus.Running || state.status === ProcessStatus.NeedsInput || state.status === "running" || state.status === "needs_input"
      )
      .map(([issueNumber]) => issueNumber);
  }

  getAvailableAgentIndex(maxConcurrent: number): number | undefined {
    const usedAgents = new Set<number>();
    for (const state of this.states.values()) {
      const isActive =
        state.status === ProcessStatus.Running ||
        state.status === ProcessStatus.NeedsInput ||
        state.status === "running" ||
        state.status === "needs_input";
      if (isActive && state.agent_index !== undefined && state.agent_index !== null) {
        usedAgents.add(state.agent_index);
      }
    }

    for (let i = 0; i < maxConcurrent; i += 1) {
      if (!usedAgents.has(i)) {
        return i;
      }
    }

    return undefined;
  }

  getAgentIssues(agentIndex: number): number[] {
    return Array.from(this.states.entries())
      .filter(([, state]) => state.agent_index === agentIndex)
      .filter(([, state]) =>
        state.status === ProcessStatus.Running || state.status === ProcessStatus.NeedsInput || state.status === "running" || state.status === "needs_input"
      )
      .map(([issueNumber]) => issueNumber);
  }

  getStateFilePath(): string {
    return this.stateFile;
  }
}
